### PR TITLE
feat: race condition guard on batch creation

### PR DESCRIPTION
**Pull request by hachall**
## Batch capacity gating — reservation ledger with fail-safe race design
**What this PR does**
Adds a reservation-ledger approach to priority window, capacity gating for batch creation. Before fusillade creates a batch, control-layer records a short-lived reservation for the expected request count. After fusillade commits, the reservation is released. This bridges the gap between "we decided there's capacity" and "the requests are actually visible in the pending-count query" (fusillade completes its transaction to write the rows to requests table).

Architecture: three-phase pipeline
```
Phase 1 — Reserve   (~1 ms, dwctl write pool)
  ├─ BEGIN tx
  ├─ pg_advisory_xact_lock per (model_id, completion_window)
  ├─ read active reservations      ← dwctl write pool, inside tx
  ├─ read pending counts           ← fusillade write pool, separate connection
  ├─ combine → capacity check
  ├─ INSERT reservation
  └─ COMMIT  (lock released, reservation visible)

Phase 2 — create_batch   (fusillade, variable duration)
  └─ atomic tx: INSERT batches → INSERT requests → COMMIT

Phase 3 — Release   (~1 ms)
  └─ UPDATE reservations SET released_at = now()
```

The advisory lock serialises Phase 1 for all concurrent callers on the same (model_id, window) pair. It is transaction-scoped, so it is automatically released when Phase 1 commits — Phase 2 runs fully outside the lock, preserving creation throughput for large batches (non blocking on expensive fusillade.create_batch). The Phase 1 window is kept as minimal as possible, as all batch creations must funnel single file through it. 

**Reservation TTL**

Reservations exist as a ledger in the db, and hold capacity while a batch is being created. If a server fails or is restarted during batch creations, these reservations would not be released and permanently reduce capacity. To avoid this, they have a time-to-live of 10mins, after which they expire. This needs to be long enough that they never expire while a batch is still being created, but not too long they dent capacity for a long time after a server restart. Note, this only affects post failure or restarts, under healthy operation the reservations will be released immediately after batch creation. Under the current configuration, the sum of all batch sizes mid creation during a restart will be the  capacity reduction for new batches coming in for the next 10 minutes (from moment of server stopping/failing, not the restart time)  - this is possibly too conservative, and can be edited in `config.reservation_ttl_secs` as desired. 

**Unavoidable race window** — and why it's fail-safe:

Pending counts are read from the fusillade write pool (a separate connection from the dwctl reservation transaction). Under PostgreSQL's default READ COMMITTED isolation, each statement holds its own point-in-time snapshot. This means the two capacity reads can observe different committed states if a concurrent batch creation finishes exactly between them.

The critical moment is the swap point: the instant when create_batch commits (requests go 0 → N in fusillade) and the reservation is released (N → 0 in dwctl). A new caller straddling this swap point can see inconsistent state across the two reads.

Read order is the mitigation. Reservations are read first, pending counts second:

```
                        swap point
                            │
Caller B reads reservations ◄─── BEFORE swap → sees 100  (reservation still present)
                            │
                    [swap fires here]
                            │
Caller B reads pending  ────►─── AFTER swap  → sees 100  (requests now committed)

B totals: 100 + 100 = 200  →  conservative double-count  →  under-accepts  ✓
```

The reversed order (read pending first, reservations second) produces the dangerous case:

```
Caller B reads pending      ◄─── BEFORE swap → sees 0   (requests not yet committed)
                            │
                    [swap fires here]
                            │
Caller B reads reservations ────►─── AFTER swap  → sees 0   (reservation released)

B totals: 0 + 0 = 0  →  system looks idle  →  over-accepts  ✗
```

The race window itself is fixed at roughly the time between two consecutive database round-trips (~a few milliseconds). It cannot be closed without either (a) reading both sources from the same connection in the same snapshot, or (b) wrapping Phase 2 inside the advisory lock — which would serialise all of [create_batch](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) end-to-end and destroy throughput. The chosen design accepts the race exists and constrains it to always fail in the safe direction.

Remaining theoretical edge case
There is a residual worst-case scenario where the scheduler suspends the caller for the entire duration of Phase 2 (potentially seconds), causing the opposite straddling:

```
B reads reservations AFTER swap  → 0   (reservation already released)
B reads pending BEFORE swap      → 0   (requests not yet committed)
B totals: 0 → could over-accept
```

This requires the OS to suspend the calling async task between two consecutive await points for the full duration of a database transaction — practically negligible, orders of magnitude less likely than the original race this PR closes. The reservation TTL (10 minutes) acts as a safety net for any leaked reservations regardless.

Status

Concern | Status
-- | --
Concurrent reservation race (the common case) | ✅ Closed — advisory lock serialises Phase 1
Swap-point race → over-acceptance | ✅ Closed — read order makes race conservative
Swap-point race → under-acceptance (at most one batch) | ⚠️ Residual — acceptable, fail-safe by design
Leaked reservations | ✅ Covered — TTL expiry as safety net
Multi-model deadlock | ✅ Prevented — deterministic UUID lock ordering

<h3>Greptile Summary</h3>

Implements a reservation-ledger system to close race conditions in batch capacity gating. Before creating a batch, the system now records short-lived reservations that hold capacity during the `create_batch` operation, bridging the gap between "capacity check passed" and "requests visible in database."

**Architecture:**
- **Phase 1 (Reserve)**: Advisory locks serialize capacity checks per (model, window), reads active reservations from dwctl, reads pending counts from fusillade, checks combined capacity, inserts reservation rows (~1ms)
- **Phase 2 (Create)**: Fusillade atomically writes batch and requests (variable duration, runs outside advisory lock)
- **Phase 3 (Release)**: Scope guard releases reservations unconditionally, whether batch creation succeeds or fails (~1ms)

**Key design decisions:**
- Advisory locks (`pg_advisory_xact_lock`) prevent concurrent reservation races, acquired in deterministic UUID order to prevent deadlocks
- Read ordering (reservations first, pending counts second) makes the unavoidable swap-point race fail-safe by double-counting rather than missing capacity
- RAII scope guards ensure reservations are released even on errors/panics
- 10-minute TTL provides safety net for leaked reservations (server crashes, runtime unavailable)
- Comprehensive test coverage for insert/release/expiry logic and capacity rejection scenarios

**Residual behavior:**
The acknowledged swap-point race can cause under-acceptance of at most one batch when reads straddle the moment a concurrent batch completes. This is acceptable fail-safe design.

<h3>Confidence Score: 4/5</h3>

- Safe to merge with minor residual race condition that fails conservatively
- Implementation is architecturally sound with advisory locks, RAII cleanup, TTL safety net, and comprehensive tests. The deliberate read ordering ensures any swap-point race errs toward under-acceptance rather than over-acceptance. Score reflects the acknowledged residual edge case (scheduler suspension between reads) which is practically negligible and documented as acceptable fail-safe behavior.
- No files require special attention - all changes are well-tested and documented

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| dwctl/migrations/066_model_reservation_ledger.sql | Adds `batch_capacity_reservations` table with optimized partial index for active reservations, clean schema design |
| dwctl/src/db/handlers/capacity_reservations.rs | New repository handler for capacity reservations with comprehensive tests covering insert, release, and expiry logic |
| dwctl/src/api/handlers/batches.rs | Implements three-phase reservation pipeline with advisory locks, scope guards for cleanup, and fail-safe race handling via read ordering |
| dwctl/src/config.rs | Adds `reservation_ttl_secs` config with validation ensuring positive values and 10-minute default, includes comprehensive tests |
| dwctl/Cargo.toml | Upgrades fusillade to v7.0.0 and adds scopeguard dependency for RAII cleanup |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client
    participant BatchHandler
    participant DwctlDB as Dwctl DB (write pool)
    participant FusilladDB as Fusillade DB (write pool)
    participant ScopeGuard

    Client->>BatchHandler: POST /batches (create batch request)
    
    Note over BatchHandler,DwctlDB: Phase 1: Reserve Capacity (~1ms)
    BatchHandler->>DwctlDB: BEGIN transaction
    
    loop For each (model_id, window) pair
        BatchHandler->>DwctlDB: pg_advisory_xact_lock(model_id, window)
        Note over DwctlDB: Serializes concurrent callers<br/>for same model+window
    end
    
    BatchHandler->>DwctlDB: SELECT active reservations<br/>(released_at IS NULL, not expired)
    DwctlDB-->>BatchHandler: Reserved counts per model
    
    BatchHandler->>FusilladDB: SELECT pending request counts<br/>(separate connection)
    FusilladDB-->>BatchHandler: Pending counts per model
    
    BatchHandler->>BatchHandler: Combine reservations + pending<br/>Check capacity vs throughput
    
    alt Insufficient Capacity
        BatchHandler->>DwctlDB: ROLLBACK
        BatchHandler-->>Client: 429 Too Many Requests
    else Has Capacity
        BatchHandler->>DwctlDB: INSERT reservations<br/>(expires_at = now + TTL)
        BatchHandler->>DwctlDB: COMMIT
        Note over DwctlDB: Advisory locks released<br/>Reservations now visible
        
        BatchHandler->>ScopeGuard: Create RAII guard<br/>(will release on scope exit)
        
        Note over BatchHandler,FusilladDB: Phase 2: Create Batch (ms-seconds)
        BatchHandler->>FusilladDB: create_batch()<br/>INSERT batches + requests
        FusilladDB-->>BatchHandler: Batch created
        
        Note over ScopeGuard,DwctlDB: Phase 3: Release (~1ms)
        ScopeGuard->>DwctlDB: UPDATE reservations<br/>SET released_at = now()
        Note over ScopeGuard: Guard runs even on errors<br/>TTL is safety net if guard fails
        
        BatchHandler-->>Client: 201 Created (batch response)
    end
```

<sub>Last reviewed commit: 0dad4d5</sub>

### CODE REVIEW

File: dwctl/src/api/handlers/sla_capacity.rs [hunk 1]
```diff
@@ -102,7 +102,7 @@ pub fn check_sla_capacity(
 /// Returns the window duration in seconds. Invalid or negative values
 /// default to 24 hours (86400 seconds). Very large values are clamped
 /// to MAX_WINDOW_SECONDS to prevent overflow in capacity calculations.
-fn parse_window_to_seconds(window: &str) -> i64 {
+pub(super) fn parse_window_to_seconds(window: &str) -> i64 {
```
Comment (pjb157) [inline dwctl/src/api/handlers/sla_capacity.rs]: Strange not seen this before

File: dwctl/migrations/066_model_reservation_ledger.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+    expires_at TIMESTAMPTZ NOT NULL,
+    released_at TIMESTAMPTZ DEFAULT NULL
+);
+
+-- Fast lookup for summing active reservations per model + window.
+-- Partial index keeps it small and hot.
+CREATE INDEX idx_batch_capacity_reservations_active
+    ON batch_capacity_reservations (model_id, completion_window, expires_at)
+    WHERE released_at IS NULL;
```
[the remark this answers is not in the document]
Comment (hachall) [reply dwctl/migrations/066_model_reservation_ledger.sql]: i thiink this is was my original design, and why i wanted to merge async batches into this workload

File: dwctl/src/api/handlers/batches.rs [hunk 3]
```diff
@@ -243,96 +241,295 @@ pub async fn create_batch<P: PoolProvider>(
...
+        &model_ids_by_alias,
+        &windows,
+        &states,
+        &model_aliases,
+    )
+    .await?;
+
+    let batch = state.request_manager.create_batch(batch_input).await;
+
+    if let Err(err) = release_capacity_reservations(&state, &reservation_ids).await {
```
[the remark this answers is not in the document]
Comment (fergusfinn) [reply dwctl/src/api/handlers/batches.rs]: i think i misunderstood greptiles comment, mine's different - i agree we want to drop the reservation regardless of error or success. It's just that the code as is doesn't necessarily do that (the only extant edge case is panics inside `create_batch`, but the more likely thing is someone refactoring the control flow to question mark the `create_batch` call, which with this code structure will skip the cleanup). 

The way to make it a bit more robust is to have the lifecycle of the 'reservation' tied to some object that is dropped when you leave this function (however you leave - whether by early return or normal return). This is usually called  RAII(`https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization`) - and you do it in rust by returning a `Reservation` struct from `reserve_capacity`, and then doing the `release_capacity` on it's `drop` method (tho like i said you can't have async drop, so maybe its too hard here).
